src/libsyntax/parse/parser.rs

```diff
@@ -2182,9 +2212,154 @@ impl<'a> Parser<'a> {
+                let (args, bindings) = match snapshot.parse_generic_args() {
+                    Ok(value) => {
+                        debug!(
+                            "parse_path_segment: (snapshot success) snapshot.count={:?}",
+                            snapshot.unmatched_angle_bracket_count,
+                        );
+                        mem::replace(self, snapshot);
+                        value
+                    },
+                    Err(mut e) => {
+                        debug!(
+                            "parse_path_segment: (snapshot failure) snapshot.count={:?}",
+                            snapshot.unmatched_angle_bracket_count,
+                        );
+                        if style == PathStyle::Expr && snapshot.unmatched_angle_bracket_count > 0 {
+                            // Cancel error from being unable to find `>`. We know the error
+                            // must have been this due to a non-zero unmatched angle bracket
+                            // count.
+                            e.cancel();
+
+                            // Eat the unmatched angle brackets.
+                            for _ in 0..snapshot.unmatched_angle_bracket_count {
+                                self.eat_lt();
+                            }
+
+                            // Make a span over ${unmatched angle bracket count} characters.
+                            let span = lo.with_hi(
+                                lo.lo() + BytePos(snapshot.unmatched_angle_bracket_count)
+                            );
+                            let plural = snapshot.unmatched_angle_bracket_count > 1;
+                            self.diagnostic()
+                                .struct_span_err(
+                                    span,
+                                    &format!(
+                                        "unmatched angle bracket{}",
+                                        if plural { "s" } else { "" }
+                                    ),
+                                )
+                                .span_suggestion_with_applicability(
+                                    span,
+                                    &format!(
+                                        "remove extra angle bracket{}",
+                                        if plural { "s" } else { "" }
+                                    ),
+                                    String::new(),
+                                    Applicability::MachineApplicable,
+                                )
+                                .emit();
+
+                            // Try again without unmatched angle bracket characters.
+                            self.parse_generic_args()?
+                        } else {
+                            mem::replace(self, snapshot);
+                            return Err(e);
+                        }
+                    },
+                };
+
```

Comment:
Could you move all this code into a separate function?

If previously `fn parse_path_segment` looked somewhat like a "reference implementation"/"executable pseudocode" so you could infer the actual language grammar from it, now most of the function is some ad hoc diagnostics having nothing in common with the language.
Change like this reduce codebase readability *so* much.

Reply:
Since every piece of ad hoc diagnostics code is a burden that slows down development and [introduces bugs](https://github.com/rust-lang/rust/issues/57849) when something new is added, we should really weight it against usefulness of the diagnostics as well.
I can't say I'm too happy with this balance in this PR.